### PR TITLE
rust: upgrade to Rust 1.66.0

### PR DESCRIPTION
**Pull request by ojeda**
Required changes:

  - Some new Clippy warnings (introduced since 1.62.0), as well as a few others that weren't triggered before, to be cleaned up or disabled as needed later on:
    + `borrow_deref_ref` (new in 1.63.0).
    + `explicit_auto_deref` (new in 1.64.0).
    + `bool_to_int_with_if` (new in 1.65.0).
    + `needless_borrow`.
    + `type_complexity`.
    + `unnecessary_cast` (this one I allowed it only on `CONFIG_ARM`).

  - `rustdoc` lint `broken_intra_doc_links` triggering on links pointing to `macro_export` `macro_rules` defined in the same module (i.e. appearing in the crate root).
    + However, even if the warning appears, the link still gets generated like in previous versions, thus it is a bit confusing -- I opened https://github.com/rust-lang/rust/issues/106142, and it appears that the link still being generated was a compatibility measure, thus we will need to adapt to the new behavior.
    + Thus I went ahead and cleaned those up in the second commit.

  - `#[const_trait]` in `RawDeviceId` (1.66.0 due to https://github.com/rust-lang/rust/pull/100982)

  - `-Aunused-imports` for `core`, fixed upstream in https://github.com/rust-lang/rust/pull/105434, and prevented for the future for that `cfg` in https://github.com/rust-lang/rust/pull/105811.

I will have to reword the first commit for upstreaming and possibly split it somehow to make it easier to read. For the moment, this should unblock those waiting for 1.66.0 and start testing it on our side.

As a triple-check, I redid the functions by hand and spotted a doc sentence missed from an old upgrade. Nothing important, but I included it here as the third commit too.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
cc: @marcan @dcavalca

**Comment by Conan-Kudo** [comment]
cc: @marcan @davide125

**Comment by aliceinwire** [comment]
Few of this problem are referenced in the following PR:
- borrow_deref_ref (new in 1.63.0) https://github.com/Rust-for-Linux/linux/pull/900
- https://github.com/Rust-for-Linux/linux/pull/898 very complex type used

File: rust/macros/module.rs [hunk 1]
```diff
@@ -143,6 +143,7 @@ fn param_ops_path(param_type: &str) -> &'static str {
     }
 }
 
+#[allow(clippy::type_complexity)]
```
Comment (aliceinwire) [inline rust/macros/module.rs:146]: Referenced in PR https://github.com/Rust-for-Linux/linux/pull/898

File: rust/kernel/driver.rs [hunk 2]
```diff
@@ -119,6 +119,7 @@ impl<T: DriverOps> Drop for Registration<T> {
 ///   - [`RawDeviceId::ZERO`] is actually a zeroed-out version of the raw device id.
 ///   - [`RawDeviceId::to_rawid`] stores `offset` in the context/data field of the raw device id so
 ///     that buses can recover the pointer to the data.
+#[const_trait]
```
Comment (bjorn3) [inline rust/kernel/driver.rs:122]: Did something break without this?

Comment (ojeda) [reply rust/kernel/driver.rs:122]: It was due to (added it to OP):

```txt
error: ~const can only be applied to `#[const_trait]` traits
   --> rust/kernel/driver.rs:156:19
    |
156 |         T: ~const RawDeviceId + Copy,
    |                   ^^^^^^^^^^^

error: const `impl` for trait `RawDeviceId` which is not marked with `#[const_trait]`
   --> rust/kernel/of.rs:40:19
    |
40  | unsafe impl const driver::RawDeviceId for DeviceId {
    |                   ^^^^^^^^^^^^^^^^^^^
    |
   ::: rust/kernel/driver.rs:122:1
    |
122 | pub unsafe trait RawDeviceId {
    | - help: mark `RawDeviceId` as const: `#[const_trait]`
    |
    = note: marking a trait with `#[const_trait]` ensures all default method bodies are `const`
    = note: adding a non-const method body in the future would be a breaking change

error: aborting due to 2 previous errors
```

**Review by bjorn3 — APPROVED** [review]

**Comment by alex** [comment]
Only did a limited review, but LGTM.

**Comment by YakoYakoYokuYoku** [comment]
Builds without warnings, LGTM

```
Reviewed-by: Martin Rodriguez Reboredo <yakoyoku@gmail.com>
Tested-by: Martin Rodriguez Reboredo <yakoyoku@gmail.com>
```

**Comment by Conan-Kudo** [comment]
cc: @asahilina

**Review by nbdd0121 — APPROVED** [review]

**Review by vincenzopalazzo — APPROVED** [review]

**Comment by ojeda** [comment]
Thanks everyone! That was awesome.

@nbdd0121 @vincenzopalazzo @bjorn3 I am transforming your GH approvals into Reviewed-bys -- let me know if you didn't intend that (in which case I will remove it when I submit it for upstream).

File: rust/macros/module.rs (hunk 1 above)
Comment (ojeda) [reply rust/macros/module.rs:146]: Thanks Alice!

**Comment by ojeda** [comment]
v2: no changes, just collected the tags.

**Comment by aliceinwire** [comment]
```
Reviewed-by: Alice Ferrazzi <alice.ferrazzi@miraclelinux.com>
Tested-by: Alice Ferrazzi <alice.ferrazzi@miraclelinux.com>
```

**Review by aliceinwire — APPROVED** [review]

**Comment by Conan-Kudo** [comment]
Running a test build now with this PR: 
* COPR: https://copr.fedorainfracloud.org/coprs/build/5183132
* Git: https://gitlab.com/fedora-asahi/kernel-asahi/-/commit/9eb7faa3606ae7ca6c57cc4d8fce1994c636eb55

**Comment by Conan-Kudo** [comment]
Running a test build now with this PR: 
* COPR: https://copr.fedorainfracloud.org/coprs/build/5183132
* Git: https://gitlab.com/fedora-asahi/kernel-asahi/-/commit/687190e3dffe204384b025810a259036cc86868f

**Review by Conan-Kudo — APPROVED** [review]
So my build is fixed, but the result seems to cause Xorg to crash. This probably means that @asahilina needs to adapt her code to Rust 1.66.

Otherwise, this looks reasonable.

```
Reviewed-by: Neal Gompa <neal@gompa.dev>
Tested-by: Neal Gompa <neal@gompa.dev>
```

**Comment by Conan-Kudo** [comment]
Also, it probably makes sense to add `Reported-by: Neal Gompa <neal@gompa.dev>` here, since this was triggered by my post in the Zulip.

**Review by Conan-Kudo — APPROVED** [review]
Actually, we apparently tested the wrong kernel build. The correct kernel build works, so I give this a total 👍🏾

```
Reported-by: Neal Gompa <neal@gompa.dev>
Reviewed-by: Neal Gompa <neal@gompa.dev>
Tested-by: Neal Gompa <neal@gompa.dev>
```

**Comment by Conan-Kudo** [comment]
I also needed a minor tweak from @ericcurtin to get bindgen working too: https://gitlab.com/fedora-asahi/kernel-asahi/-/commit/7407fdf7716f3233da56a5fb4e2e28fffb775fa5

**Comment by ojeda** [comment]
Thanks a lot for the extra reviews & tests, @aliceinwire and @Conan-Kudo, very much appreciated.

v3: tags collected for the forthcoming merge and added a proper commit message for the first patch.

> Also, it probably makes sense to add Reported-by: Neal Gompa <neal@gompa.dev> here, since this was triggered by my post in the Zulip.

`Reported-by` is meant for bugs only according to the docs. In any case, version upgrades will keep happening until we reach a stable release, and others (e.g. @bjorn3 if I recall correctly) suggested/discussed the 1.66 upgrade in the past. For non-routine version upgrades, we could perhaps do `Suggested-by`s, but really they make sense only after we are in a stable situation.

**Comment by Conan-Kudo** [comment]
🤷🏾‍♂️

**Comment by ojeda** [comment]
GitHub's UI seems bugged on my side... It has merged the branch, but the PR remains open, and the UI here claims there are conflicts.

**Comment by Conan-Kudo** [comment]
did you do it manually or via the button?

**Comment by ojeda** [comment]
Via the button (we don't push on this branch). I have been experiencing GitHub timeouts very often in the last few days, too (e.g. some minutes ago when clicking the "Compare" button above for v2..v3).

**Pull request closed without merging**